docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+# Artifact Outputs for Incremental Generators
+
+## Summary
+
+This proposal adds a mechanism that allows [incremental generators](incremental-generators.md) to output non-source artifacts. This would allow for text or binary files to be emitted to disk alongside the compiled assembly.
+
+### Rationale
+
+There exists a desire for incremental source generators be able to write out a set of accompanying files, alongside the generated sources that they add to the compilation. An example of this is JavaScript interop with .NET code. When exposing a .NET method to JavaScript today it is possible to generate the C# interop source code that enables the call to be made from JavaScript, but the user must still hand author the corresponding JavaScript code. Instead, the incremental generator could emit the required C# code alongside a JavaScript file that provides the equivalent JavaScript interop code.
+
+### High Level Design Goals
+
+- Allow Incremental Generators to emit text or binary files to disk alongside the emitted .NET assembly
+- The user can control where the generated artifacts are emitted with a suitable default
+- Be optimal in terms of file writes / copies and ensure we keep the correct incremental build semantics for MSBuild
+- Do not have to be run as part of the IDE experience. Files are only emitted to disk as part of command line compilation
```

Comment:
For the scenario I'm interested in (outputting OpenAPI files) there is a desire to produce the files as part of the IDE experience so they can integrate into things like VS's API Explorer. Is the primary concern here around perf?

Reply:
Yes, allowing arbitrary writes to the disk during IDE will be a horrible perf cliff, I don't think there is any way we could do it efficiently.

That being said, I wonder if we could enable it to run at least to the point of providing a filename. That way we could (in theory) allow the IDE to list the artifacts being produced and potentially run the generator for that artifact if the user wants to view it?